File: samples/language-impl-comparison/haskell/sql_parser.hs
{-# LANGUAGE LambdaCase #-}

-- 簡易SQL Parser
-- SELECT, WHERE, JOIN, ORDER BY など基本的な構文のみ対応

module Main where

import Control.Applicative
import Control.Monad
import Data.Char (isAlphaNum, isDigit, isLetter, isSpace, toLower)
import Data.List (intercalate)

-- AST定義

data Query = SelectQuery
  { columns :: [Column]
  , from :: TableRef
  , whereClause :: Maybe Expr
  , joins :: [Join]
  , orderBy :: Maybe [(Expr, OrderDirection)]
  }
  deriving (Show)

data Column
  = AllColumns
  | ColumnExpr Expr (Maybe String)
  deriving (Show)

data TableRef = TableRef
  { table :: String
  , tableAlias :: Maybe String
  }
  deriving (Show)

data Join = Join
  { joinType :: JoinType
  , joinTable :: TableRef
  , onCondition :: Expr
  }
  deriving (Show)

data JoinType
  = InnerJoin
  | LeftJoin
  | RightJoin
  | FullJoin
  deriving (Show)

data OrderDirection
  = Asc
  | Desc
  deriving (Show, Eq)

data Expr
  = LiteralExpr Literal
  | ColumnRef String
  | QualifiedColumn String String
  | BinaryOp BinOp Expr Expr
  | UnaryOp UnOp Expr
  | FunctionCall String [Expr]
  | Parenthesized Expr
  deriving (Show)

data Literal
  = IntLit Int
  | StringLit String
  | BoolLit Bool
  | NullLit
  deriving (Show)

data BinOp
  = Add | Sub | Mul | Div | Mod
  | Eq | Ne | Lt | Le | Gt | Ge
  | And | Or | Like
  deriving (Show)

data UnOp
  = Not
  | IsNull
  | IsNotNull
  deriving (Show)

-- パーサーコンビネーター

newtype Parser a = Parser { runParser :: String -> Maybe (a, String) }

instance Functor Parser where
  fmap f (Parser p) = Parser $ \input ->
    case p input of
      Just (result, rest) -> Just (f result, rest)
      Nothing -> Nothing

instance Applicative Parser where
  pure x = Parser $ \input -> Just (x, input)
  Parser pf <*> Parser px = Parser $ \input ->
    case pf input of
      Just (f, rest) ->
        case px rest of
          Just (x, rest') -> Just (f x, rest')
          Nothing -> Nothing
      Nothing -> Nothing

instance Monad Parser where
  return = pure
  Parser p >>= f = Parser $ \input ->
    case p input of
      Just (result, rest) -> runParser (f result) rest
      Nothing -> Nothing

instance Alternative Parser where
  empty = Parser $ const Nothing
  Parser p1 <|> Parser p2 = Parser $ \input ->
    case p1 input of
      Just result -> Just result
      Nothing -> p2 input

satisfy :: (Char -> Bool) -> Parser Char
satisfy pred = Parser $ \case
  (c:cs) | pred c -> Just (c, cs)
  _ -> Nothing

char :: Char -> Parser Char
char c = satisfy (== c)

string :: String -> Parser String
string str = Parser $ \input ->
  if take (length str) input == str
    then Just (str, drop (length str) input)
    else Nothing

skipWhitespace :: Parser ()
skipWhitespace = Parser $ \input ->
  Just ((), dropWhile isSpace input)

lexeme :: Parser a -> Parser a
lexeme p = skipWhitespace *> p <* skipWhitespace

symbol :: String -> Parser ()
symbol str = lexeme (string str) *> pure ()

keyword :: String -> Parser ()
keyword kw = lexeme $ do
  skipWhitespace
  _ <- Parser $ \input ->
    let len = length kw
        slice = take len input
    in if map toLower slice == map toLower kw
       then let rest = drop len input
            in case rest of
                 (c:_) | isAlphaNum c -> Nothing
                 _ -> Just ((), rest)
       else Nothing
  skipWhitespace
  pure ()

reservedWords :: [String]
reservedWords =
  [ "select", "from", "where", "join", "inner", "left", "right", "full"
  , "on", "and", "or", "not", "like", "order", "by", "asc", "desc"
  , "null", "true", "false", "as", "is"
  ]

identifier :: Parser String
identifier = lexeme $ do
  skipWhitespace
  first <- satisfy (\c -> isLetter c || c == '_')
  rest <- many (satisfy (\c -> isAlphaNum c || c == '_'))
  let name = first : rest
  if map toLower name `elem` reservedWords
    then empty
    else do
      skipWhitespace
      pure name

integer :: Parser Int
integer = lexeme $ do
  skipWhitespace
  digits <- some (satisfy isDigit)
  skipWhitespace
  pure (read digits)

stringLiteral :: Parser String
stringLiteral = lexeme $ do
  skipWhitespace
  _ <- char '\''
  content <- many (satisfy (/= '\''))
  _ <- char '\''
  skipWhitespace
  pure content

literalParser :: Parser Literal
literalParser =
  (keyword "null" *> pure NullLit)
    <|> (keyword "true" *> pure (BoolLit True))
    <|> (keyword "false" *> pure (BoolLit False))
    <|> (IntLit <$> integer)
    <|> (StringLit <$> stringLiteral)

-- 式パーサー（演算子優先度対応）

exprParser :: Parser Expr
exprParser = orExpr

orExpr :: Parser Expr
orExpr = do
  left <- andExpr
  orExprCont left

orExprCont :: Expr -> Parser Expr
orExprCont left =
  (do
      keyword "or"
      right <- andExpr
      orExprCont (BinaryOp Or left right))
    <|> pure left

andExpr :: Parser Expr
andExpr = do
  left <- comparisonExpr
  andExprCont left

andExprCont :: Expr -> Parser Expr
andExprCont left =
  (do
      keyword "and"
      right <- comparisonExpr
      andExprCont (BinaryOp And left right))
    <|> pure left

comparisonExpr :: Parser Expr
comparisonExpr = do
  left <- additiveExpr
  (symbol "<=" *> (BinaryOp Le left <$> additiveExpr))
    <|> (symbol ">=" *> (BinaryOp Ge left <$> additiveExpr))
    <|> (symbol "<>" *> (BinaryOp Ne left <$> additiveExpr))
    <|> (symbol "!=" *> (BinaryOp Ne left <$> additiveExpr))
    <|> (symbol "=" *> (BinaryOp Eq left <$> additiveExpr))
    <|> (symbol "<" *> (BinaryOp Lt left <$> additiveExpr))
    <|> (symbol ">" *> (BinaryOp Gt left <$> additiveExpr))
    <|> (keyword "like" *> (BinaryOp Like left <$> additiveExpr))
    <|> pure left

additiveExpr :: Parser Expr
additiveExpr = do
  left <- multiplicativeExpr
  additiveExprCont left

additiveExprCont :: Expr -> Parser Expr
additiveExprCont left =
  (do
      symbol "+"
      right <- multiplicativeExpr
      additiveExprCont (BinaryOp Add left right))
    <|> (do
            symbol "-"
            right <- multiplicativeExpr
            additiveExprCont (BinaryOp Sub left right))
    <|> pure left

multiplicativeExpr :: Parser Expr
multiplicativeExpr = do
  left <- postfixExpr
  multiplicativeExprCont left

multiplicativeExprCont :: Expr -> Parser Expr
multiplicativeExprCont left =
  (do
      symbol "*"
      right <- postfixExpr
      multiplicativeExprCont (BinaryOp Mul left right))
    <|> (do
            symbol "/"
            right <- postfixExpr
            multiplicativeExprCont (BinaryOp Div left right))
    <|> (do
            symbol "%"
            right <- postfixExpr
            multiplicativeExprCont (BinaryOp Mod left right))
    <|> pure left

postfixExpr :: Parser Expr
postfixExpr = do
  expr <- unaryExpr
  postfixExprCont expr

postfixExprCont :: Expr -> Parser Expr
postfixExprCont expr =
  (do
      keyword "is"
      (do
          keyword "not"
          keyword "null"
          postfixExprCont (UnaryOp IsNotNull expr))
        <|> (do
                keyword "null"
                postfixExprCont (UnaryOp IsNull expr)))
    <|> pure expr

unaryExpr :: Parser Expr
unaryExpr =
  (keyword "not" *> (UnaryOp Not <$> unaryExpr))
    <|> atomExpr

atomExpr :: Parser Expr
atomExpr =
  (do
      symbol "("
      expr <- exprParser
      symbol ")"
      pure (Parenthesized expr))
    <|> (do
            name <- identifier
            (do
                symbol "("
                args <- functionArgs
                symbol ")"
                pure (FunctionCall name args))
              <|> (do
                      symbol "."
                      col <- identifier
                      pure (QualifiedColumn name col))
              <|> pure (ColumnRef name))
    <|> (LiteralExpr <$> literalParser)

functionArgs :: Parser [Expr]
functionArgs =
  (do
      first <- exprParser
      rest <- many (symbol "," *> exprParser)
      pure (first : rest))
    <|> pure []

-- クエリパーサー

columnListParser :: Parser [Column]
columnListParser =
  (symbol "*" *> pure [AllColumns])
    <|> (do
            first <- columnExprParser
            rest <- many (symbol "," *> columnExprParser)
            pure (first : rest))

columnExprParser :: Parser Column
columnExprParser = do
  expr <- exprParser
  alias <- optional ((keyword "as" *> identifier) <|> identifier)
  pure (ColumnExpr expr alias)

tableRefParser :: Parser TableRef
tableRefParser = do
  tbl <- identifier
  alias <- optional ((keyword "as" *> identifier) <|> identifier)
  pure (TableRef tbl alias)

joinTypeParser :: Parser JoinType
joinTypeParser =
  (keyword "inner" *> keyword "join" *> pure InnerJoin)
    <|> (keyword "left" *> keyword "join" *> pure LeftJoin)
    <|> (keyword "right" *> keyword "join" *> pure RightJoin)
    <|> (keyword "full" *> keyword "join" *> pure FullJoin)
    <|> (keyword "join" *> pure InnerJoin)

joinParser :: Parser Join
joinParser = do
  jt <- joinTypeParser
  tbl <- tableRefParser
  keyword "on"
  condition <- exprParser
  pure (Join jt tbl condition)

orderByItemParser :: Parser (Expr, OrderDirection)
orderByItemParser = do
  expr <- exprParser
  dir <- (keyword "asc" *> pure Asc)
         <|> (keyword "desc" *> pure Desc)
         <|> pure Asc
  pure (expr, dir)

orderByParser :: Parser [(Expr, OrderDirection)]
orderByParser = do
  keyword "order"
  keyword "by"
  first <- orderByItemParser
  rest <- many (symbol "," *> orderByItemParser)
  pure (first : rest)

selectQueryParser :: Parser Query
selectQueryParser = do
  keyword "select"
  cols <- columnListParser
  keyword "from"
  tbl <- tableRefParser
  js <- many joinParser
  wc <- optional (keyword "where" *> exprParser)
  ob <- optional orderByParser
  pure (SelectQuery cols tbl wc js ob)

parseSQL :: String -> Maybe Query
parseSQL input =
  case runParser parser input of
    Just (result, rest)
      | all isSpace rest -> Just result
    _ -> Nothing
  where
    parser = skipWhitespace *> selectQueryParser <* skipWhitespace <* optional (symbol ";") <* skipWhitespace

-- レンダリング（検証用）

renderQuery :: Query -> String
renderQuery (SelectQuery cols tbl wc js ob) =
  "SELECT " ++ colsStr ++ " " ++ fromStr ++ " " ++ joinsStr ++ whereStr ++ orderStr
  where
    colsStr = intercalate ", " (map renderColumn cols)
    fromStr = "FROM " ++ table tbl ++ maybe "" (" AS " ++) (tableAlias tbl)
    joinsStr = unwords (map renderJoin js)
    whereStr = maybe "" (\e -> " WHERE " ++ renderExpr e) wc
    orderStr = maybe "" renderOrderBy ob

renderColumn :: Column -> String
renderColumn AllColumns = "*"
renderColumn (ColumnExpr expr alias) =
  renderExpr expr ++ maybe "" (" AS " ++) alias

renderJoin :: Join -> String
renderJoin (Join jt tbl cond) =
  joinTypeStr ++ " " ++ table tbl ++ " ON " ++ renderExpr cond
  where
    joinTypeStr = case jt of
      InnerJoin -> "INNER JOIN"
      LeftJoin -> "LEFT JOIN"
      RightJoin -> "RIGHT JOIN"
      FullJoin -> "FULL JOIN"

renderExpr :: Expr -> String
renderExpr (LiteralExpr lit) = renderLiteral lit
renderExpr (ColumnRef name) = name
renderExpr (QualifiedColumn tbl col) = tbl ++ "." ++ col
renderExpr (BinaryOp op left right) =
  "(" ++ renderExpr left ++ " " ++ renderBinOp op ++ " " ++ renderExpr right ++ ")"
renderExpr (UnaryOp Not e) = "NOT " ++ renderExpr e
renderExpr (UnaryOp IsNull e) = renderExpr e ++ " IS NULL"
renderExpr (UnaryOp IsNotNull e) = renderExpr e ++ " IS NOT NULL"
renderExpr (FunctionCall name args) =
  name ++ "(" ++ intercalate ", " (map renderExpr args) ++ ")"
renderExpr (Parenthesized e) = "(" ++ renderExpr e ++ ")"

renderLiteral :: Literal -> String
renderLiteral (IntLit n) = show n
renderLiteral (StringLit s) = "'" ++ s ++ "'"
renderLiteral (BoolLit b) = if b then "TRUE" else "FALSE"
renderLiteral NullLit = "NULL"

renderBinOp :: BinOp -> String
renderBinOp Add = "+"
renderBinOp Sub = "-"
renderBinOp Mul = "*"
renderBinOp Div = "/"
renderBinOp Mod = "%"
renderBinOp Eq = "="
renderBinOp Ne = "<>"
renderBinOp Lt = "<"
renderBinOp Le = "<="
renderBinOp Gt = ">"
renderBinOp Ge = ">="
renderBinOp And = "AND"
renderBinOp Or = "OR"
renderBinOp Like = "LIKE"

renderOrderBy :: [(Expr, OrderDirection)] -> String
renderOrderBy items =
  " ORDER BY " ++ intercalate ", " (map renderItem items)
  where
    renderItem (expr, dir) =
      renderExpr expr ++ " " ++ (if dir == Asc then "ASC" else "DESC")

-- テスト

main :: IO ()
main = do
  let testCases =
        [ "SELECT * FROM users"
        , "SELECT name, age FROM users WHERE age > 18"
        , "SELECT u.name, o.total FROM users u INNER JOIN orders o ON u.id = o.user_id"
        , "SELECT name FROM users WHERE active = true ORDER BY name ASC"
        ]

  putStrLn "=== SQL Parser Test ==="
  mapM_ testCase testCases
  where
    testCase sql = do
      putStrLn $ "\nInput: " ++ sql
      case parseSQL sql of
        Just query -> do
          putStrLn "Parsed: OK"
          putStrLn $ "Rendered: " ++ renderQuery query
        Nothing ->
          putStrLn "Error: Parse failed"

File: samples/language-impl-comparison/ocaml/sql_parser.ml
(* 簡易SQL Parser - OCaml実装 *)
(* SELECT, WHERE, JOIN, ORDER BY対応 *)
(* Angstromパーサーコンビネーターライブラリを使用 *)

open Angstrom

(* AST定義 *)
type order_direction = Asc | Desc

type join_type =
  | InnerJoin
  | LeftJoin
  | RightJoin
  | FullJoin

type bin_op =
  | Add | Sub | Mul | Div | Mod
  | Eq | Ne | Lt | Le | Gt | Ge
  | And | Or | Like

type un_op =
  | Not
  | IsNull
  | IsNotNull

type literal =
  | IntLit of int
  | FloatLit of float
  | StringLit of string
  | BoolLit of bool
  | NullLit

type expr =
  | Literal of literal
  | Column of string
  | QualifiedColumn of string * string
  | BinaryOp of bin_op * expr * expr
  | UnaryOp of un_op * expr
  | FunctionCall of string * expr list
  | Parenthesized of expr

type column =
  | AllColumns
  | ColumnExpr of expr * string option

type table_ref = {
  table : string;
  alias : string option;
}

type join = {
  join_type : join_type;
  join_table : table_ref;
  on_condition : expr;
}

type order_by = {
  order_columns : (expr * order_direction) list;
}

type query = {
  columns : column list;
  from_table : table_ref;
  where_clause : expr option;
  joins : join list;
  order_by : order_by option;
}

(* パーサー補助関数 *)
let is_whitespace = function
  | ' ' | '\t' | '\n' | '\r' -> true
  | _ -> false

let whitespace = take_while is_whitespace

let line_comment =
  string "--" *> take_till (fun c -> c = '\n') *> return ()

let block_comment =
  string "/*" *> take_till (fun c -> false) *> string "*/" *> return ()
  (* 簡略化: ネストなしのブロックコメント *)

let sc = skip_many (whitespace <|> line_comment <|> block_comment)

let lexeme p = p <* sc

let symbol s = lexeme (string s)

(* キーワード（大文字小文字を区別しない） *)
let keyword kw =
  lexeme (
    string_ci kw >>= fun _ ->
    peek_char >>= function
    | Some c when (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
                  (c >= '0' && c <= '9') || c = '_' ->
        fail "keyword followed by alphanumeric"
    | _ -> return ()
  )

(* 識別子 *)
let is_ident_start c =
  (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c = '_'

let is_ident_cont c =
  is_ident_start c || (c >= '0' && c <= '9')

let identifier =
  let reserved = ["select"; "from"; "where"; "join"; "inner"; "left";
                  "right"; "full"; "on"; "and"; "or"; "not"; "like";
                  "order"; "by"; "asc"; "desc"; "null"; "true"; "false"; "as"] in
  lexeme (
    satisfy is_ident_start >>= fun first ->
    take_while is_ident_cont >>= fun rest ->
    let name = String.make 1 first ^ rest in
    let lower = String.lowercase_ascii name in
    if List.mem lower reserved then
      fail ("reserved word: " ^ name)
    else
      return name
  )

(* リテラル *)
let integer =
  lexeme (take_while1 (fun c -> c >= '0' && c <= '9')) >>= fun s ->
  return (IntLit (int_of_string s))

let float_lit =
  lexeme (
    take_while1 (fun c -> c >= '0' && c <= '9') >>= fun int_part ->
    char '.' >>= fun _ ->
    take_while1 (fun c -> c >= '0' && c <= '9') >>= fun frac_part ->
    return (FloatLit (float_of_string (int_part ^ "." ^ frac_part)))
  )

let string_lit =
  lexeme (
    char '\'' *> take_till (fun c -> c = '\'') <* char '\'' >>= fun s ->
    return (StringLit s)
  )

let literal =
  choice [
    keyword "null" *> return NullLit;
    keyword "true" *> return (BoolLit true);
    keyword "false" *> return (BoolLit false);
    float_lit;
    integer;
    string_lit;
  ]

(* 式パーサー（演算子優先度を考慮した再帰下降パーサー） *)
let rec expr_parser () =
  or_expr ()

and or_expr () =
  and_expr () >>= fun left ->
  many (keyword "or" *> and_expr ()) >>= fun rights ->
  return (List.fold_left (fun acc r -> BinaryOp (Or, acc, r)) left rights)

and and_expr () =
  cmp_expr () >>= fun left ->
  many (keyword "and" *> cmp_expr ()) >>= fun rights ->
  return (List.fold_left (fun acc r -> BinaryOp (And, acc, r)) left rights)

and cmp_expr () =
  add_expr () >>= fun left ->
  option None (
    choice [
      symbol "=" *> return Eq;
      symbol "<>" *> return Ne;
      symbol "!=" *> return Ne;
      symbol "<=" *> return Le;
      symbol ">=" *> return Ge;
      symbol "<" *> return Lt;
      symbol ">" *> return Gt;
      keyword "like" *> return Like;
    ] >>= fun op ->
    add_expr () >>= fun right ->
    return (Some (op, right))
  ) >>= function
  | None -> return left
  | Some (op, right) -> return (BinaryOp (op, left, right))

and add_expr () =
  mul_expr () >>= fun left ->
  many (
    choice [symbol "+" *> return Add; symbol "-" *> return Sub] >>= fun op ->
    mul_expr () >>= fun right ->
    return (op, right)
  ) >>= fun ops ->
  return (List.fold_left (fun acc (op, r) -> BinaryOp (op, acc, r)) left ops)

and mul_expr () =
  unary_expr () >>= fun left ->
  many (
    choice [
      symbol "*" *> return Mul;
      symbol "/" *> return Div;
      symbol "%" *> return Mod;
    ] >>= fun op ->
    unary_expr () >>= fun right ->
    return (op, right)
  ) >>= fun ops ->
  return (List.fold_left (fun acc (op, r) -> BinaryOp (op, acc, r)) left ops)

and unary_expr () =
  choice [
    keyword "not" *> unary_expr () >>= fun e ->
    return (UnaryOp (Not, e));
    postfix_expr ();
  ]

and postfix_expr () =
  primary_expr () >>= fun e ->
  option None (
    keyword "is" *>
    option false (keyword "not" *> return true) >>= fun is_not ->
    keyword "null" *>
    return (if is_not then IsNotNull else IsNull)
  ) >>= function
  | None -> return e
  | Some op -> return (UnaryOp (op, e))

and primary_expr () =
  choice [
    symbol "(" *> expr_parser () <* symbol ")" >>= fun e ->
    return (Parenthesized e);
    function_call ();
    column_ref ();
    literal >>= fun lit -> return (Literal lit);
  ]

and function_call () =
  identifier >>= fun name ->
  symbol "(" *>
  sep_by (symbol ",") (expr_parser ()) <* symbol ")" >>= fun args ->
  return (FunctionCall (name, args))

and column_ref () =
  identifier >>= fun first ->
  option None (symbol "." *> identifier >>= fun col -> return (Some col)) >>= function
  | None -> return (Column first)
  | Some col -> return (QualifiedColumn (first, col))

(* カラムリスト *)
let column_list =
  choice [
    symbol "*" *> return [AllColumns];
    sep_by1 (symbol ",") (
      expr_parser () >>= fun e ->
      option None (option () (keyword "as") *> identifier >>= fun alias -> return (Some alias)) >>= fun alias ->
      return (ColumnExpr (e, alias))
    );
  ]

(* テーブル参照 *)
let table_ref =
  identifier >>= fun table ->
  option None (option () (keyword "as") *> identifier >>= fun alias -> return (Some alias)) >>= fun alias ->
  return { table; alias }

(* JOIN句 *)
let join_type =
  choice [
    keyword "inner" *> keyword "join" *> return InnerJoin;
    keyword "left" *> keyword "join" *> return LeftJoin;
    keyword "right" *> keyword "join" *> return RightJoin;
    keyword "full" *> keyword "join" *> return FullJoin;
    keyword "join" *> return InnerJoin;
  ]

let join_clause =
  join_type >>= fun jt ->
  table_ref >>= fun tbl ->
  keyword "on" *> expr_parser () >>= fun cond ->
  return { join_type = jt; join_table = tbl; on_condition = cond }

(* ORDER BY句 *)
let order_by_clause =
  keyword "order" *> keyword "by" *>
  sep_by1 (symbol ",") (
    expr_parser () >>= fun e ->
    option Asc (
      choice [
        keyword "asc" *> return Asc;
        keyword "desc" *> return Desc;
      ]
    ) >>= fun dir ->
    return (e, dir)
  ) >>= fun cols ->
  return { order_columns = cols }

(* SELECT文 *)
let select_query =
  keyword "select" *> column_list >>= fun cols ->
  keyword "from" *> table_ref >>= fun from ->
  many join_clause >>= fun joins ->
  option None (keyword "where" *> expr_parser () >>= fun e -> return (Some e)) >>= fun where_c ->
  option None (order_by_clause >>= fun ob -> return (Some ob)) >>= fun order ->
  return {
    columns = cols;
    from_table = from;
    where_clause = where_c;
    joins = joins;
    order_by = order;
  }

(* パブリックAPI *)
let parse input =
  parse_string ~consume:All (sc *> select_query <* option () (symbol ";")) input

(* レンダリング関数 *)
let render_literal = function
  | IntLit n -> string_of_int n
  | FloatLit f -> string_of_float f
  | StringLit s -> "'" ^ s ^ "'"
  | BoolLit b -> if b then "TRUE" else "FALSE"
  | NullLit -> "NULL"

let render_binop = function
  | Add -> "+" | Sub -> "-" | Mul -> "*" | Div -> "/" | Mod -> "%"
  | Eq -> "=" | Ne -> "<>" | Lt -> "<" | Le -> "<=" | Gt -> ">" | Ge -> ">="
  | And -> "AND" | Or -> "OR" | Like -> "LIKE"

let rec render_expr = function
  | Literal lit -> render_literal lit
  | Column name -> name
  | QualifiedColumn (tbl, col) -> tbl ^ "." ^ col
  | BinaryOp (op, left, right) ->
      "(" ^ render_expr left ^ " " ^ render_binop op ^ " " ^ render_expr right ^ ")"
  | UnaryOp (Not, e) -> "NOT " ^ render_expr e
  | UnaryOp (IsNull, e) -> render_expr e ^ " IS NULL"
  | UnaryOp (IsNotNull, e) -> render_expr e ^ " IS NOT NULL"
  | FunctionCall (name, args) ->
      name ^ "(" ^ String.concat ", " (List.map render_expr args) ^ ")"
  | Parenthesized e -> "(" ^ render_expr e ^ ")"

let render_column = function
  | AllColumns -> "*"
  | ColumnExpr (e, None) -> render_expr e
  | ColumnExpr (e, Some alias) -> render_expr e ^ " AS " ^ alias

let render_query q =
  let cols = String.concat ", " (List.map render_column q.columns) in
  let from = "FROM " ^ q.from_table.table ^
    (match q.from_table.alias with None -> "" | Some a -> " AS " ^ a) in
  let joins = String.concat " " (List.map (fun j ->
    let jt = match j.join_type with
      | InnerJoin -> "INNER JOIN"
      | LeftJoin -> "LEFT JOIN"
      | RightJoin -> "RIGHT JOIN"
      | FullJoin -> "FULL JOIN" in
    jt ^ " " ^ j.join_table.table ^ " ON " ^ render_expr j.on_condition
  ) q.joins) in
  let where = match q.where_clause with
    | None -> ""
    | Some e -> " WHERE " ^ render_expr e in
  let order = match q.order_by with
    | None -> ""
    | Some ob ->
        let cols = String.concat ", " (List.map (fun (e, dir) ->
          render_expr e ^ " " ^ (if dir = Asc then "ASC" else "DESC")
        ) ob.order_columns) in
        " ORDER BY " ^ cols in
  "SELECT " ^ cols ^ " " ^ from ^ " " ^ joins ^ where ^ order

(* テスト *)
let () =
  print_endline "=== OCaml SQL Parser テスト ===";
  let test_sql = "SELECT * FROM users WHERE id = 1" in
  match parse test_sql with
  | Ok q ->
      print_endline ("パース成功: " ^ test_sql);
      print_endline ("レンダリング: " ^ render_query q)
  | Error msg ->
      print_endline ("パースエラー: " ^ msg)